### DISCUSSION TITLE
Webpack Plugin re-processing all assets on watch mode

### DISCUSSION THREAD

**Discussion by stormojm** [webpack/webpack#16651] [Q&A] [answered]
I have a webpack plugin which generates a variant css file for every css file in bundle (based off mini-css-extract-plugin).  In webpack 4, I used the compilation `additionalAssets` hook.  In webpack 5, I migrated that to `processAssets`:

```
compiler.hooks.thisCompilation.tap(pluginName, (compilation) => {
  compilation.hooks.processAssets.tapPromise({
    name: pluginName,
    stage: Compilation.PROCESS_ASSETS_STAGE_ADDITIONAL,
  }, (assets) => {
    // filter css assets and create an extra asset file based on it.
    // eventually calling compilation.emitAsset
    return this.createVariantFile(assets);
  });
});
```

The issue I am having is this processAsset gets called on every re-build when in watch mode (using webpack-dev-server).  Since assets is every asset and not just the changed files it can take a fair amount of time.  What I need is someway to only generate the new asset if the original was changed.

Is there any way to find out which assets are being changed and only process them?  Or do I have to rely on a webpack caching object to skip files that I already have a variant for?

**Comment by alexander-akait** [comment, marked as answer]
You can't know when files will be changed when not, so the best solution is cache, look at code here https://github.com/webpack-contrib/html-minimizer-webpack-plugin/blob/master/src/index.js#L398 you can use the same approach in your plugin, so you will have:
1. No extra work in the watch mode
2. Cache between compilation (if you will have filesystem cache your plugins will be faster between rebuild)

The code is easy to read and integrate, so feel free to feedback and questions

**Reply by stormojm** [reply]
Thanks, thats the general approach I was taking, but wanted to make sure it was correct.
